text/0567-use-and-resources.md

```diff
@@ -0,0 +1,953 @@
+was used for that same key will be used again, so a different value can be
+returned instead.
+
+#### Creating vs Updating
+
+Using initializers with `@use` is very nice for end users, but presents a
+problem for authors of `ResourceManager`s - initializers don't receive
+arguments, so there is no direct way to change behavior between creating a
+resource and updating it. Even if they could, it would also be very unergonomic
+to force users to pass around arguments through their code:
+
+```js
+@use myState = (owner, isCreating) => isCreating ?
+  // creating Resource, return a new instance
+  Resource.create(owner, this.args.foo) :
+  // updating Resource, use the previous instance
+  null
+```
+
+Instead, `@use` will be accompanied by a helper function: `getUseInfo`.
+
+```ts
+declare function getUseInfo(): {
+  owner: Owner;
+  isCreating: boolean;
+  isUpdating: boolean;
+};
+```
+
+This information will be populated before calling the initializer, and can be
+called _exactly once_ during initialization. Calling a second time will throw an
+assertion, which is meant to prevent abuse of the API by other code (e.g. other
+classes that are initialized when creating the Resource). For safety, it should
+always be called immediately in `DEBUG` mode by any Resource implementation,
+even if it is not actually needed. Not calling it at all in `DEBUG` mode will
+also throw an assertion to help make sure implementers don't leak it
+accidentally to user code.
+
+#### Setting Arguments
+
+Additionally, initializers can only return one value, which is the manager
+handle itself. An important aspect of resources is that they should be able to
+respond to incoming arguments that are consumed while running the initializer,
+and in order to do this, the arguments have to be passed to the manager. In
+order to do this, users can use the `setUseArgs` method.
+
+```ts
+declare function setUseArgs(...args: unknown[]): void;
+```
+
+Like with `getUseInfo`, this function can be used _exactly once_ during the
+initializer's execution to save off the arguments, and not using it in DEBUG
+mode will throw an assertion to help make sure it isn't accidentally leaked to
+user code.
+
+#### Scheduling `@use` Initialization
+
+As mentioned above, the timing semantics of `@use` properties are that they
+start automatically, asynchronously, after the parent class they are defined on
+with is created, or when they are first consumed, whichever is first. Because of
+limitations of Ember's current stage 1/legacy decorators, this is not possible
+just by decorating a class element such as a field or method, since we intend to
+create a getter and not a field. Getters do not run any code on initialization.
+
+As such, Ember will provide a function that can be used to schedule starting any
+`@use` instances defined on the class, `scheduleUseStart()`
+
+```js
+class GlimmerComponent {
+  constructor() {
+    scheduleUseStart(this);
```

Comment:
We should probably discuss this in a bit more depth in the RFC.

Within Glimmer, there is already a hierarchy of destroyables - components, helpers, and modifiers generally. This hierarchy was designed with an RAII like model - essentially, children associate themselves with their parent as rendered, and if the parent is destroyed, then so are the children, automatically.

Resources would leverage this structure, associating themselves with the nearest parent, and destroying when it is destroyed. For non-framework cases, like if a resource is defined on a utility class, this would be the component where the utility class is used/instantiated.

I think that the only case this doesn't really cover is utility classes that have their own lifecycle, independent of the parent. We could consider adding another utility function that exposes the ability to teardown a class directly, since we're exposing `scheduleUseStart` as is.

Reply:
I believe the latest draft is much more clear about how the association would work, via the `useResource()` method. If you still think this could use some fleshing out, let me know.